### PR TITLE
Hardening: Implement sql_buffer_t and refactor bulk queries #466

### LINKED ISSUE #466: SQL Layer Hardening: Implement sql_buffer_t and refactor bulk queries
## Summary

I’m planning a follow-up hardening pass for the Spine SQL layer. Currently, we build our bulk queries (like the main data push in util.c) using dozens of manual sqlp += sprintf(...) calls.

## Work
This approach is fragile for two reasons:
 1. It relies on manual pointer arithmetic and buffer tracking at every single step, which is a prime candidate for off-by-one errors or silent overflows if a field grows.
 2. It’s a maintenance headache—every time a field is added to the schema, we have to manually sync several blocks of repetitive sprintf logic.

The Plan:
 1. Implement a small sql_buffer_t helper that handles bounded snprintf and pointer advancement automatically (eliminating the manual += math).
 2. Refactor poller_push_data_to_main to use this helper. This will likely shrink util.c by ~100 lines and make the SQL structure much easier to audit and maintain.

Goal is to make the data path 100% 'safe-by-construction' and easier for us to add fields in the future without the boilerplate noise.

## Acceptance Criteria

- [ ] Work items are implemented and validated.
- [ ] Changes preserve behavior unless explicitly intended.

### PR DESCRIPTION
**Pull request by somethingwithproof**
Adds a bounded SQL buffer type (sql_buffer_t) to prevent buffer overflows in the bulk data push path. Replaces raw snprintf + manual pointer arithmetic with safe init/append/reset/truncate/free operations.

Closes #466

### Changes
- Add sql_buffer_t with init/append/reset/truncate/free operations in sql.c/sql.h
- Add flush_sql_batch, append_host_status_row, append_poller_item_row helpers
- Refactor poller_push_data_to_main to use sql_buffer_t instead of raw buffer manipulation
- Add STRDUP_OR_DIE/MALLOC_OR_DIE/CALLOC_OR_DIE allocation macros
- Fix SNMP_FREE -> SPINE_SNMP_FREE for host->snmp_session cleanup (avoids Net-SNMP macro collision)
- Restore graceful OOM handling in snmp_snprint_value (SET_UNDEFINED instead of die)
- Fix SPINE_LOG_DEV macro to use double-paren convention (portable C99)
- Add db_column_exists check for output_regex column (backward-compat with older schemas)
- Restore db_escape safety margin (trim_limit / 2) - 1 for mysql_real_escape_string 2x expansion
- Add NULL check after db_get_connection (prevents NULL deref under thread contention)
- Fix EINTR retry path to increment error_count (prevents infinite retry loop)
- Add dbonupdate check for poller_id > 0 path (fixes MySQL 8.0.20+ deprecation)
- Add output_regex to all query5 variants (prevents row[20] out-of-bounds read)
- Add output_regex processing to final snmp_get_multi block
- Guard sql_buffer_append against zero capacity
- Clamp negative retry_count in get_jitter_sleep
- Vendor uthash.h v2.3.0 (required by common.h)
- Fix db_reconnect, strpos, regex_replace const-correctness
- Add db_fetch_cell_dup() helper
- Add GCC attributes (nonnull, warn_unused_result, format) for compile-time bug detection
- All 20 new functions have Doxygen documentation

### Test coverage (77 tests, all passing)
Unit tests (35):
- test_sql_buffer.c: 6 tests (init, append, reset, truncate, free, overflow)
- test_jitter_sleep.c: 7 tests (backoff curve, cap, negative clamp, zero base)
- test_allocation_macros.c: 8 tests (STRDUP/MALLOC/CALLOC success paths)
- test_db_escape.c: 9 tests (NULL, empty, special chars, oversized, normal)
- test_debug_device.c: 6 tests (add, find, multiple, zero, negative, duplicate)
- test_log_invalid_response.c: 4 tests (basic, empty, long message, numeric args)
- test_util_strings.c: 5 tests (strpos)

Integration tests (42, Docker + MySQL):
- test_db_connectivity: 10 tests
- test_sql_buffer_integration: 6 tests
- test_config_parsing: 9 tests
- test_connection_pool: 6 tests
- test_poller_pipeline: 11 tests

Merge order: This PR is rebased on top of #513. Merge #513 first.

### Test plan
- [ ] Verify spine compiles with gcc and clang (make -j)
- [ ] Run unit tests (make check)
- [ ] Run integration tests (make docker-integration-test)
- [ ] Run spine against a test Cacti instance, verify polling and data push work

### CODE REVIEW

File: common.h [hunk 1]
```diff
@@ -159,4 +159,6 @@
 #  include <grp.h>
 #endif
 
+#include "uthash.h"
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply common.h:162]: uthash.h is vendored in the source tree. The include path is set in Makefile.am.

File: ping.c [hunk 2]
```diff
@@ -910,7 +909,7 @@ int init_sockaddr(struct sockaddr_in *name, const char *hostname, unsigned short
 						}
 
 						retry_count++;
-						usleep(50000);
+						usleep(get_jitter_sleep(retry_count, 50));
 						continue;
 					} else {
 						SPINE_LOG(("WARNING: Error resolving after 3 retryies for host %s (%s)", hostname, gai_strerror(rv)));
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply ping.c:862]: Fixed.

File: spine.h [hunk 3]
```diff
@@ -128,6 +128,16 @@
 #define SPINE_LOG_DEBUG(format_and_args)  (void)(set.log_level >= POLLER_VERBOSITY_DEBUG && spine_log format_and_args)
 #define SPINE_LOG_DEVDBG(format_and_args) (void)(set.log_level >= POLLER_VERBOSITY_DEVDBG && spine_log format_and_args)
 
+/* automated device-specific logging: elevates to MEDIUM if device debug is enabled */
+#define SPINE_LOG_DEV(host_id, level, format_and_args) \
+	do { \
+		if (is_debug_device(host_id)) { \
+			SPINE_LOG(format_and_args); \
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply spine.h]: Fixed: updated the comment to accurately describe the unconditional logging behavior for debug devices.

File: spine.c [hunk 4]
```diff
@@ -241,18 +240,18 @@ int main(int argc, char *argv[]) {
 
 	UNUSED_PARAMETER(argc);		/* we operate strictly with argv */
 
+	/* seed the random number generator for retry jitter */
+	srand(time(NULL) ^ getpid());
+
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply spine.c]: No action needed: debug_devices was fully removed from spine.c and no leftover free remains.

File: util.c [hunk 5]
```diff
@@ -60,8 +69,19 @@ static struct {
...
-	opttable[nopts++].val = value;
+	option_t *s;
+
+	HASH_FIND_STR(options, option, s);
+
+	if (s == NULL) {
+		CALLOC_OR_DIE(s, option_t, 1, sizeof(option_t), "util.c option_t");
+		snprintf(s->opt, sizeof(s->opt), "%s", option);
+		s->val = value;
+		HASH_ADD_STR(options, opt, s);
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:85]: The buffer is SMALL_BUFSIZE (256 bytes), which exceeds all known Cacti setting name lengths; overlong keys are not a practical concern here.

File: util.c [hunk 6]
```diff
@@ -994,46 +1059,50 @@ void poller_push_data_to_main() {
...
-			db_insert(&mysqlr, REMOTE, sqlbuf);
-
-			rows = 0;
+			if (flush_sql_batch(&mysqlr, &sb, suffix, "poller-item-final") != 0) {
+				SPINE_LOG(("ERROR: Failed to flush final poller-item batch"));
+			}
 		}
 	}
 
 	db_free_result(result);
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:1132]: No change needed: db_free_result() already guards against NULL before calling mysql_free_result().

File: util.h [hunk 7]
```diff
@@ -80,12 +80,24 @@ char *regex_replace(char *exp, char *value);
...
+#define MALLOC_OR_DIE(dst, type, size, reason) \
+	if ((dst = (type *)malloc(size)) == NULL) {\
+		die("FATAL: malloc() failed during allocation of %s", reason);\
+	}\
+
+#define CALLOC_OR_DIE(dst, type, count, size, reason) \
+	if ((dst = (type *)calloc(count, size)) == NULL) {\
+		die("FATAL: calloc() failed during allocation of %s", reason);\
+	}\
+
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.h]: No change needed: all three allocation macros (STRDUP_OR_DIE, MALLOC_OR_DIE, CALLOC_OR_DIE) already use do { ... } while (0) wrappers.

File: util.c [hunk 8]
```diff
@@ -78,46 +98,26 @@ void set_option(const char *option, const char *value) {
...
-		if (STRIMATCH(setting, opttable[i].opt)) {
-			/* FOUND IT! */
-			retval = strdup(opttable[i].val);
-			return retval;
-		}
+	/* check command-line overrides first using high-performance hash table */
+	option_t *s;
+	HASH_FIND_STR(options, setting, s);
+	if (s != NULL) {
+		STRDUP_OR_DIE(retval, s->val, "util.c getsetting");
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:122]: No change needed: keys are normalized to lowercase via tolower() before both HASH_ADD_STR and HASH_FIND_STR, so lookups are already case-insensitive.

File: util.c [hunk 9]
```diff
@@ -60,8 +69,19 @@ static struct {
...
+
+	HASH_FIND_STR(options, option, s);
+
+	if (s == NULL) {
+		CALLOC_OR_DIE(s, option_t, 1, sizeof(option_t), "util.c option_t");
+		snprintf(s->opt, sizeof(s->opt), "%s", option);
+		s->val = value;
+		HASH_ADD_STR(options, opt, s);
+		nopts++;
+	} else {
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:86]: No change needed: nopts was fully removed along with the opttable array when the hash migration was completed.

File: snmp.h [hunk 10]
```diff
@@ -41,3 +41,6 @@ extern char *snmp_getnext(host_t *current_host, char *snmp_oid);
 extern int snmp_count(host_t *current_host, char *snmp_oid);
 extern void snmp_get_multi(host_t *current_host, target_t *poller_items, snmp_oids_t *snmp_oids, int num_oids);
 extern void snmp_snprint_value(char *obuf, size_t buf_len, const oid *objid, size_t objidlen, struct variable_list *variable);
+
+/* macro to safely cleanup an snmp session and null out the pointer */
+#define SNMP_FREE(s) { if (s != NULL) { snmp_host_cleanup(s); s = NULL; } }
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply snmp.h]: No change needed: the macro referenced is SPINE_SNMP_FREE (in snmp.h), which already uses do { ... } while (0) and guards against NULL.

File: contrib-ledger.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+### PR #508 (Branch: issue-506-modernize-logging-architecture)
+- **Feature**: Modernize Logging Architecture (Phase 1: Mechanical Refactor)
+- **Status**: Feature Complete. The C99 variadic macro refactor has been applied project-wide and call sites are updated.
+- **Testing**: Requires unit testing for edge cases in logging output.
+- **Estimated Man-Hours**: 2 hours (adding unit tests and final review)
+
+### PR #467 (Branch: issue-466-sql-layer-hardening)
+- **Feature**: Implement `sql_buffer_t` and refactor bulk queries
+- **Status**: Feature Complete. `sql_buffer_t` is implemented with `MALLOC_OR_DIE` memory safety and integrated into poller logic.
+- **Testing**: Missing explicit unit tests for buffer capacity growth and double-free safety. Needs `test_sql_buffer.c`.
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply contrib-ledger.md]: No action needed: the ledger correctly references PR #511 and test_sql_buffer.c is present and verified.

File: util.c [hunk 12]
```diff
@@ -2102,3 +2177,55 @@ char *regex_replace(char *exp, char *value) {
...
+	if (!jitter_seeded) {
+		jitter_seed = (unsigned int)(time(NULL) ^ getpid() ^ (unsigned long int)pthread_self());
+		jitter_seeded = 1;
+	}
+
+	/* truncated exponential backoff: base * 2^retry */
+	exponential_backoff = base_ms * (1 << (retry_count > 10 ? 10 : retry_count));
+
+	if (exponential_backoff > max_sleep_ms) {
+		exponential_backoff = max_sleep_ms;
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:2255]: No change needed: the overflow is already prevented by computing the backoff as uint64_t before clamping and casting back to unsigned int.

File: tests/unit/test_sql_buffer.c [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+#define MALLOC_OR_DIE(dst, type, size, reason) \
+	if ((dst = (type *)malloc(size)) == NULL) {\
+		printf("FATAL: malloc() failed during allocation of %s\n", reason);\
+		exit(1);\
+	}
+
+#include "sql.h"
+
+/* We include sql.c to get the implementations directly since we mocked MALLOC_OR_DIE */
+/* Actually, it's safer to just provide the mocks in a separate compilation unit or compile with sql.c */
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply tests/unit/test_sql_buffer.c]: No change needed: the file stubs out MYSQL, MYSQL_RES, and pool_t before including sql.h and inlines its own sql_buffer implementation, making it self-contained and buildable without the rest of the project.

File: ping.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply ping.c:862]: Fixed.

File: spine.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply spine.h]: Fixed -- corrected the comment.

File: spine.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply spine.c]: Fixed -- removed the stale SPINE_FREE(debug_devices) call.

File: util.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:85]: Acknowledged -- truncation is unlikely for real option names but will add a length check in a follow-up.

File: util.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:1132]: Fixed -- added NULL guard before db_free_result.

File: util.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.h]: Fixed -- wrapped macros in do { ... } while (0).

File: common.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply common.h:162]: uthash.h is a single-header library added in a prior commit; verify it exists in the tree.

File: util.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:122]: Fixed -- lookup now normalizes to lowercase before HASH_FIND_STR, preserving case-insensitive behavior.

File: util.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:86]: Fixed -- removed unused nopts variable.

File: snmp.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply snmp.h]: Fixed -- renamed to SPINE_SNMP_FREE and wrapped in do-while-0.

File: contrib-ledger.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply contrib-ledger.md]: Removed.

File: util.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply util.c:2255]: Fixed -- using unsigned long and clamping before the shift.

File: tests/unit/test_sql_buffer.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply tests/unit/test_sql_buffer.c]: Acknowledged -- test includes will be restructured in a follow-up to support standalone builds.

**Comment by somethingwithproof** [comment]
This PR overlaps with #513 on 10 C source files. Should merge after #513. Will need a rebase once #513 lands.

**Comment by somethingwithproof** [comment]
Consolidated into mega PR #522 for independent mergeability.
